syspage.c

```diff
@@ -0,0 +1,187 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system kernel
+ *
+ * Syspage
+ *
+ * Copyright 2021 Phoenix Systems
+ * Authors: Hubert Buczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "lib/lib.h"
+#include "syspage.h"
+
+
+struct {
+	syspage_t *syspage;
+} syspage_common;
+
+
+size_t syspage_mapSize(void)
+{
+	size_t nb = 0;
+	const syspage_map_t *map = syspage_common.syspage->maps;
+
+	if (map == NULL)
+		return nb;
+
+	do {
+		++nb;
+	} while ((map = map->next) != syspage_common.syspage->maps);
+
+	return nb;
+}
+
+
+const syspage_map_t *syspage_mapList(void)
+{
+	return syspage_common.syspage->maps;
+}
+
+
+const syspage_map_t *syspage_mapDataResolve(unsigned int type, const void *data)
+{
+	const syspage_map_t *map = syspage_common.syspage->maps;
+
+	if (map == NULL)
+		return NULL;
+
+	do {
+		switch (type) {
+			case syspage_idResolve:
+				if (*((unsigned int *)data) == map->id)
+					return map;
+				break;
+			case syspage_addrResolve:
+				if ((addr_t)data < map->end && (addr_t)data >= map->start)
+					return map;
+				break;
+
+			case syspage_nameResolve:
+				if (hal_strcmp((const char *)data, map->name) == 0)
+					return map;
+				break;
+
+			default:
+				return NULL;
```

Comment:
I think that API would be easier to use and more efficient if there were 3 funtions for 3 types of resolve:
```
const syspage_map_t *syspage_mapIdResolve(unsigned int id);
const syspage_map_t *syspage_mapAddrResolve(addr_t addr);
const syspage_map_t *syspage_mapNameResolve(const char *name);
```

Right now this API is non-intuitive and increases number of MISRA violations.

Reply:
Initially I wrote the same API as you proposed. 
However, taking into account main assumptions from writing API in `plo/hal.h`, that we should limit the count of functions I decided to concatenate 3 functions into one.
If from your perspective it is non-intuitive, I can change it.

Reply:
I think it is, I cannot use these function by just reading prototypes - I need to read it's body as well. Also, this approach introduces many casts (enum -> int -> enum, const void * -> unsigned int */ addr_t / const char *) that violate MISRA rules.

Reply:
Done